src/main/java/com/google/step/snippet/servlets/FeedbackServlet.java

```diff
@@ -0,0 +1,177 @@
+public class FeedbackServlet extends HttpServlet {
+
+  private static final String UP = "upvote";
+  private static final String DOWN = "downvote";
+  private static final String TOTAL = "totalvotes";
+  private static final String UPVOTED = "upvoted";
+  private static final String DOWNVOTED = "downvoted";
+  private static final String URL = "url";
+  private static final String FEEDBACK = "feedback";
+  private static final String USER = "user";
+  private static final String UID = "uid";
+  private static final String EMAIL = "email";
+  private static final String ACTIVE = "active";
+  private static final String DISABLED = "disabled";
+  private static final String TOG_UP = "toggleUpvote";
+  private static final String TOG_DOWN = "toggleDownvote";
+
+  private Entity getUserEntity(DatastoreService datastore, UserService userService) {
+    String uid = userService.getCurrentUser().getUserId();
+    Query.FilterPredicate filterUser = new Query.FilterPredicate(UID, FilterOperator.EQUAL, uid);
+    Query queryUser = new Query(USER).setFilter(filterUser);
+    Entity userEntity = datastore.prepare(queryUser).asSingleEntity();
+    if (userEntity == null) {
+      userEntity = new Entity(USER);
+      userEntity.setProperty(UID, uid);
+      userEntity.setProperty(EMAIL, userService.getCurrentUser().getEmail());
+    }
+    return userEntity;
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    JSONObject json = new JSONObject();
+    UserService userService = UserServiceFactory.getUserService();
+
+    if (userService.isUserLoggedIn()) {
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      String url = request.getParameter(URL);
+      Entity userEntity = getUserEntity(datastore, userService);
+
+      // If a user has previously voted on this card, retrieve their past voting status
+      if (url != null) {
+        ArrayList<String> userUpCards = (ArrayList<String>) userEntity.getProperty(UPVOTED);
+        ArrayList<String> userDownCards = (ArrayList<String>) userEntity.getProperty(DOWNVOTED);
+        if (userUpCards != null && userUpCards.contains(url)) {
+          json.put(TOG_UP, ACTIVE);
+        } else if (userDownCards != null && userDownCards.contains(url)) {
+          json.put(TOG_DOWN, ACTIVE);
+        }
+      }
+    }
+    response.setContentType("application/json");
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    // Invalid POST request if URL is invalid or user is not logged
+    String url = request.getParameter(URL);
+    UserService userService = UserServiceFactory.getUserService();
+    if (url == null || !userService.isUserLoggedIn()) {
+      JSONObject json = new JSONObject();
+      response.setContentType("application/json");
+      response.getWriter().println(json);
```

Comment:
i think adding a notice asking the user to log in will be better than returning an empty json?

Reply:
I can add a message popup on the javascript end when it finds the empty json to ask the user to login